_posts/2019-09-14-polynomial-multiplication.md

```diff
@@ -0,0 +1,243 @@
+---
+layout: post
+title:  "How to efficiently multiply polynomials"
+date:   2018-09-13 08:58:09 +0100
+categories: jekyll update
+---
+
+* TOC
+{:toc}
+
+## Motivation
+Many problems can be reduced to polynomial multiplication problem.
+Most familiar example is the number multiplication.
+Have you ever wondered why are you able to multiply 100000-digit numbers in Python in seconds?
+
+School algorithm is very simple, but also very slow. It's time complexity is $O(n^2)$.
+There are much more efficient algorithms and the goal of this article is to explore them.
+
+After formalizing the problem, we'll see Karatsuba algorithm for multiplying polynomials in $O(n^{1.58})$ .
+Next, we'll try to multiply polynomials using polynomial interpolation which will be the base for $O(n \log n)$ algorithms that utilize Fast Fourirer Transform (FFT) and Number Theoretic Transform (NTT).
+
+All these algorithms are [Divide and Conquer](https://en.wikipedia.org/wiki/Divide_and_conquer_algorithms) algorithms, so this article can also serve as an introduction (through examples) with that problem solving approach.
+
+## Problem definition
+In the rest of the article we'll use following ways to note polynomial of degree $n$ (all are considered equal):
+
+$p = p(x) = \sum\limits_{j=0}^n p_jx^j = p_0 + p_1x + p_2x^2 + .. + p_nx^n$
+
+At the input we have two polynomials $a$ and $b$ of same degree $n$, and the output is a polynomial $c$ of degree $2n$:
+
+$c(x) = a(x)b(x) = \sum\limits_{i=0}^n \sum\limits_{j=0}^n a_ib_j x^{i+j}$
+
+School algorithm written in C is simple and unambiguosly illustrates the problem:
+{% highlight c %}
+
+  void mul(int* a, int* b, int* c, int n) {
+    for (int i = 0; i <= 2*n; ++i) {
+      c[i] = 0;
+    }
+    for (int i = 0; i <= n; ++i) {
+      for (int j = 0; j <= n; ++j) {
+        c[i+j] += a[i] * b[j];
+      }
+    }
+  }
+{% endhighlight %}
+
+
+
+
+## Karatsuba algorithm
+We will solve the problem recursively:
+* base case is $n = 0$, then simply $c = a_0b_0$
+* otherwise ($n > 0$):
+	* split the input polynomials into halves of (almost) equal sizes:
```

Comment:
"into two factors of  (almost) equal degrees?". "Halve" i "size" su mi zbunjujuci.